### DISCUSSION TITLE
Background kept when making cylinder

### DISCUSSION THREAD

**Discussion by luco** [ImageMagick/ImageMagick#5112] [Help] [answered]
I'm trying to make a cylinder wrap via ImageMagick. It's working fine on creating the cylinder, but it's keeping the original image on back.

```
magick \
-verbose \
\( \
      -size 2585x1181 \
      xc:blue \
      \( \
         ./images/test.png \
         -geometry +820+0 \
      \) \
      -composite \
      -geometry 730x390 \
      -roll -230+0 \
      -crop 270x408+0+0 \
      +append \
\) \
-gravity south -background Transparent -splice 0x40 \
\( +clone -sparse-color barycentric '0,0 black 273,0 white' \) \
\( -clone 1 -function arcsin 0.5 \) \
\( -clone 1 -level 0%,100% \
            -function polynomial -4,4,0 -gamma 2 \
            +level 50%,0%  \) \
-delete 1 \
-virtual-pixel Transparent \
-define compose:args=0x40 \
-compose Displace \
-composite \
./output/out.png
```

![image](https://user-images.githubusercontent.com/246373/166393858-41571bc3-a59b-4890-b322-df785c281b8b.png)

How can I remove the background?

**Comment by fmw42** [comment]
Anthony's code seems to work fine.

```
convert rose: -background none -gravity south -splice 0x16 \
\( +clone -sparse-color barycentric '0,0 black 69,0 white' \) \
\( -clone 1 -function arcsin 0.5 \) \
\( -clone 1 -level 25%,75% \
-function polynomial -4,4,0 -gamma 2 \
+level 50%,0 \) \
-delete 1 \
-virtual-pixel none  -define compose:args=17x7 \
-compose Displace  -composite   rose_cylinder.png

```
![rose_cylinder](https://user-images.githubusercontent.com/13458218/166400487-4b6ac280-aed5-4212-9b83-854650ec289f.png)

**Reply by fmw42** [reply]
Post your input image so that we can test your code to see how it differs.

**Reply by luco** [reply]
![test](https://user-images.githubusercontent.com/246373/166401734-bee80e24-0842-4b2d-8afe-37eb49344657.png)

**Reply by luco** [reply]
@fmw42 There you go.

**Reply by luco** [reply]
> Anthony's code seems to work fine.
> 
> ```
> convert rose: -background none -gravity south -splice 0x16 \
> \( +clone -sparse-color barycentric '0,0 black 69,0 white' \) \
> \( -clone 1 -function arcsin 0.5 \) \
> \( -clone 1 -level 25%,75% \
> -function polynomial -4,4,0 -gamma 2 \
> +level 50%,0 \) \
> -delete 1 \
> -virtual-pixel none  -define compose:args=17x7 \
> -compose Displace  -composite   rose_cylinder.png
> ```
> 
> ![rose_cylinder](https://user-images.githubusercontent.com/13458218/166400487-4b6ac280-aed5-4212-9b83-854650ec289f.png)

Using this code the bottom gets flat as well for me.

**Comment by snibgo** [comment]
IM v7 "magick" treats alpha diferently to IM v6 "convert". I suggest inserting "-alpha off" after "-gravity south -background Transparent -splice 0x40 \". But I'm not sure what result is required.

I also suggest inserting "+repage" after "+append".

In general, when debugging complex commands like this, I suggest inserting operations like "+write x1.png" at various points in the code. Are those debugging images as you expect? This narrows down where the problem occurs.

**Reply by luco** [reply]
Thanks for replying. None of the commands took effect.
Also, about the debugging tip, it seems it's working properly, but it's "forgetting" removing the original layer.

**Comment by snibgo** [comment, marked as answer]
I don't know what result you want. You can make a mock-up in Gimp.

At "+append", there is only one image, so "+append" has no effect. I have removed it.

Perhaps this will help:
```
magick \
\( \
      -size 2585x1181 \
      xc:blue \
      \( \
         tribldg.png \
         -geometry +820+0 \
      \) \
      -composite \
      -geometry 730x390 \
      -roll -230+0 \
      -crop 270x408+0+0 +repage \
\) \
-gravity south -background None -splice 0x40 \
\( +clone -sparse-color barycentric '0,0 black 273,0 white' \) \
\( -clone 1 -alpha off -level 0%,100% \
            -function polynomial -4,4,0 -gamma 2 \
            +level 50%,0%  \) \
-delete 1 \
-virtual-pixel Transparent \
-define compose:args=0x40 \
-compose Displace \
-composite \
out.png
```
![out](https://user-images.githubusercontent.com/33812368/166551622-e760da7f-e903-4a7d-b950-cef3f8e21a04.png)
The command makes the same result in v6.9.9-50 "convert" and v7.1.0-32 "magick".

**Reply by luco** [reply]
I'm not sure why but your code works properly both on IM6 and 7

**Comment by luco** [comment]
I found why:
`\( -clone 1 -function arcsin 0.5 \) \` is leaving a copy on background.

**Reply by fmw42** [reply]
Yes, I discovered that also.  Snibgo's code does not warp into a cylinder. That is, it does not distort the image to have more distortion at the sides. It only pushes the center down more than the sides.  

One needs the arcsin distortion in the horizontal to properly show the cylindrical distortion. But I am missing something as you are correct in that it is leaving a background copy. Perhaps there is an extra image in the command line that I am not seeing with all your preprocessing at the top. There should be only 3 images for the -compose displace -composite. The first should be the input image, the second the arcsin distortion and the third the vertical displacement that is more in the center than the sides.

Also your distortion arguments of 0x40 is suspicious, since 0 will turn off the arcsin distortion and leave the image behind. That is likely what is happening. You are effectively only using the input and vertical displacement and not the arcsin.  Perhaps it is just your arguments are wrong.

**Comment by snibgo** [comment]
I removed ...
```
\( -clone 1 -function arcsin 0.5 \)
```
... because the displace args are 0x40, so the arcsin image does nothing useful. "-level 0%,100%" also does nothing useful.

If you want horizontal displacement as well as vertical, I suggest:
```
magick \
\( \
      -size 2585x1181 \
      xc:blue \
      \( \
         tribldg.png \
         -geometry +820+0 \
      \) \
      -composite \
      -geometry 730x390 \
      -roll -230+0 \
      -crop 270x408+0+0 \
      +append +repage \
\) \
-gravity south -background Transparent -splice 0x40 \
\( +clone \
   -set colorspace Gray \
   -alpha off \
   -sparse-color barycentric '0,0 black 273,0 white' \
  \( -clone 0 -function arcsin 0.5 \) \
  \( -clone 0 \
     -function polynomial -4,4,0 -gamma 2 \
     +level 50%,0%  \) \
  \( +clone \) \
  -delete 0 \
  -combine \
  -alpha off \
\) \
-virtual-pixel Transparent \
-define compose:args=20x40 \
-compose Displace \
-composite \
+channel \
out5.png
```

![out5](https://user-images.githubusercontent.com/33812368/166580924-d00c614f-4465-40f3-b124-3f55e3076f1f.png)
This also makes the same result in v6.9.9-50 "convert" and v7.1.0-32 "magick".

**Comment by fmw42** [comment]
One does not have to do the combine step and can leave it as 3 images in principle. However, it works only for IM 6 and not for IM 7 or I have not figured out what to change regarding colorspace or other IM 7 issue. So perhaps there is a bug. Note your +append and -level 0x100% are not needed.

This works for IM 6


```
convert \
-verbose \
\( \
      -size 2585x1181 \
      xc:blue \
      \( \
         test.png \
         -geometry +820+0 \
      \) \
      -composite \
      -resize 730x390 \
      -roll -230+0 \
      -crop 270x408+0+0 +repage \
\) \
-gravity south -background none -splice 0x40 \
\( -clone 0 -set colorspace Gray -alpha off -sparse-color barycentric '0,0 black 270,0 white' \) \
\( -clone 1 -function arcsin 0.5 \) \
\( -clone 1 \
            -function polynomial -4,4,0 -gamma 2 \
            +level 50%,0% \) \
-delete 1 \
-background none -virtual-pixel none \
-define compose:args=20x40 \
-compose Displace \
-composite \
test_out1.png

```

![test_out1](https://user-images.githubusercontent.com/13458218/166585959-8ff79ca4-a76d-4556-b2bd-b743f631096d.png)

But the same fails for IM 7

```
magick \
-verbose \
\( \
      -size 2585x1181 \
      xc:blue \
      \( \
         test.png \
         -geometry +820+0 \
      \) \
      -composite \
      -resize 730x390 \
      -roll -230+0 \
      -crop 270x408+0+0 +repage \
\) \
-gravity south -background none -splice 0x40 \
\( -clone 0 -set colorspace Gray -alpha off -sparse-color barycentric '0,0 black 270,0 white' \) \
\( -clone 1 -function arcsin 0.5 \) \
\( -clone 1 \
            -function polynomial -4,4,0 -gamma 2 \
            +level 50%,0% \) \
-delete 1 \
-background none -virtual-pixel none \
-define compose:args=20x40 \
-compose Displace \
-composite \
test_out2.png

```
![test_out2](https://user-images.githubusercontent.com/13458218/166586006-c71c85d4-d3e7-4965-9bde-a5ca3b7fb90b.png)

**Comment by fmw42** [comment]
If I change -set colorspace gray to -set colorspace sRGB, it looks better but the background original image is still there.

```
magick \
-verbose \
\( \
      -size 2585x1181 \
      xc:blue \
      \( \
         test.png \
         -geometry +820+0 \
      \) \
      -composite \
      -resize 730x390 \
      -roll -230+0 \
      -crop 270x408+0+0 +repage \
\) \
-gravity south -background none -splice 0x40 \
\( -clone 0 -set colorspace sRGB -alpha off -sparse-color barycentric '0,0 black 270,0 white' \) \
\( -clone 1 -function arcsin 0.5 \) \
\( -clone 1 \
            -function polynomial -4,4,0 -gamma 2 \
            +level 50%,0% \) \
-delete 1 \
-background none -virtual-pixel none \
-define compose:args=20x40 \
-compose Displace \
-composite \
test_out2b.png

```

![test_out2b](https://user-images.githubusercontent.com/13458218/166586292-21ad36d2-61de-4aa8-a3cb-b569b53dc54b.png)

Perhaps the -function arcsin is not working correctly in IM 7. Further tests will be needed.

**Reply by fmw42** [reply]
I tested the -function arcsin 0.5 with both IM 6 and IM 7 and they generate the same results. So perhaps it is the 3 image -compose displace -composite that does not work, but works when the displacement images are combined.

**Comment by snibgo** [comment]
fmw42 wrote:
> So perhaps it is the 3 image -compose displace -composite that does not work, but works when the displacement images are combined.

I suspect that is true. Looking at the source code in composite.c function CompositeImage(), for DisplaceCompositeOp and DistortCompositeOp it takes the x and y displacements from the red and green channels of a single image.

I can't find any code that takes the displacements from two separate images.

**Reply by fmw42** [reply]
Perhaps an oversight in the porting to IM 7. I do not see anything mentioned about it in the IM porting web page.

**Comment by snibgo** [comment]
I should mention that I inserted "-set colorspace Gray" for performance, so the sparse-color and other operations work on only one channel instead of three channels.

Every row of the displacement map is the same. So a better optimisation creates just one row, and scales it up to the full height.

The following command doesn't work in v6, because we can't do...
```
  -scale "%[WW]x%[HH]!"
```

```
magick \
\( \
      -size 2585x1181 \
      xc:blue \
      \( \
         tribldg.png \
         -geometry +820+0 \
      \) \
      -composite \
      -geometry 730x390 \
      -roll -230+0 \
      -crop 270x408+0+0 \
      +append +repage \
\) \
-gravity south -background Transparent -splice 0x40 \
-set option:WW %w \
-set option:HH %h \
\( -size %[WW]x1 gradient:Black-White \
  -channel R -function arcsin 0.5 \
  -channel G -function polynomial -4,4,0 -gamma 2 \
     +level 50%,0% \
  +channel \
  -scale "%[WW]x%[HH]!" \
  -alpha off \
\) \
-virtual-pixel Transparent \
-define compose:args=20x40 \
-compose Displace \
-composite \
+channel \
out6.png
```

**Comment by urban-warrior** [comment]
Thanks for the problem report.  We can reproduce it and will have a patch to fix it in the GIT main branch @ https://github.com/ImageMagick/ImageMagick later today.  The patch will be available in the beta releases of ImageMagick @ https://imagemagick.org/download/beta/ by sometime tomorrow.

**Reply by fmw42** [reply]
@urban-warrior. Is this fix going to allow separate X and Y displacement map images in IM 7 in addition to the combined single image displacement map where red is X and green is Y and blue is arbitrary and ignored that currently works?

**Comment by urban-warrior** [comment]
We simply fixed the bug so that IMv7 results match that of IMv6.  If there are other issues or enhancements, provide details here and we will review.

**Reply by fmw42** [reply]
This was that to which I was referring (copy of what I wrote above)

One does not have to do the combine step and can leave it as 3 images in principle. However, it works only for IM 6 and not for IM 7 or I have not figured out what to change regarding colorspace or other IM 7 issue. So perhaps there is a bug. Note your +append and -level 0x100% are not needed.

This works for IM 6


```
convert \
-verbose \
\( \
      -size 2585x1181 \
      xc:blue \
      \( \
         test.png \
         -geometry +820+0 \
      \) \
      -composite \
      -resize 730x390 \
      -roll -230+0 \
      -crop 270x408+0+0 +repage \
\) \
-gravity south -background none -splice 0x40 \
\( -clone 0 -set colorspace Gray -alpha off -sparse-color barycentric '0,0 black 270,0 white' \) \
\( -clone 1 -function arcsin 0.5 \) \
\( -clone 1 \
            -function polynomial -4,4,0 -gamma 2 \
            +level 50%,0% \) \
-delete 1 \
-background none -virtual-pixel none \
-define compose:args=20x40 \
-compose Displace \
-composite \
test_out1.png

```

![test_out1](https://user-images.githubusercontent.com/13458218/166585959-8ff79ca4-a76d-4556-b2bd-b743f631096d.png)

But the same fails for IM 7

```
magick \
-verbose \
\( \
      -size 2585x1181 \
      xc:blue \
      \( \
         test.png \
         -geometry +820+0 \
      \) \
      -composite \
      -resize 730x390 \
      -roll -230+0 \
      -crop 270x408+0+0 +repage \
\) \
-gravity south -background none -splice 0x40 \
\( -clone 0 -set colorspace Gray -alpha off -sparse-color barycentric '0,0 black 270,0 white' \) \
\( -clone 1 -function arcsin 0.5 \) \
\( -clone 1 \
            -function polynomial -4,4,0 -gamma 2 \
            +level 50%,0% \) \
-delete 1 \
-background none -virtual-pixel none \
-define compose:args=20x40 \
-compose Displace \
-composite \
test_out2.png

```
![test_out2](https://user-images.githubusercontent.com/13458218/166586006-c71c85d4-d3e7-4965-9bde-a5ca3b7fb90b.png)

**Comment by urban-warrior** [comment]
Recall ImageMagick supports variable channels.  To get the expected results, you need to convert two of your images from gray (1-channel)  to true color (4-channel).  Add `-type TrueColorAlpha` right after the `-delete 1` command-line argument to get expected results.   We could auto-promote images from gray to true color if compositing-- but that might violate the "user is king" tenant where some users may want to just composite the single channel gray scale image.

**Reply by fmw42** [reply]
That does not work.

```
magick \
-verbose \
\( \
      -size 2585x1181 \
      xc:blue \
      \( \
         test.png \
         -geometry +820+0 \
      \) \
      -composite \
      -resize 730x390 \
      -roll -230+0 \
      -crop 270x408+0+0 +repage \
\) \
-gravity south -background none -splice 0x40 \
\( -clone 0 -set colorspace Gray -alpha off -sparse-color barycentric '0,0 black 270,0 white' \) \
\( -clone 1 -function arcsin 0.5 \) \
\( -clone 1 \
            -function polynomial -4,4,0 -gamma 2 \
            +level 50%,0% \) \
-delete 1 \
-type truecolor \
-background none -virtual-pixel none \
-define compose:args=20x40 \
-compose Displace \
-composite \
test_out2.png

```
![test_out2](https://user-images.githubusercontent.com/13458218/167233010-928d42e4-31ea-4e19-97aa-65a8f12b0dc2.png)

It should look like this from IM 6:

![test_out1](https://user-images.githubusercontent.com/13458218/167233036-7850a1fc-e723-44dc-8dac-46a8e81ea2f7.png)

**I think the issue is that IM 7 is not allowing for 3 image displacement and requires the two displacement maps to be made into the red and green channels of a combined image so that there is only 2 images in the process (that does work as shown further up this post).**

It looks to me like the x map is not doing anything but putting the input image down, then the y map is applied again to the input and overlaid on the previous x map result.

**Comment by urban-warrior** [comment]
Your command works for us.  You may not be using the latest Beta.

**Comment by fmw42** [comment]
Using the latest IM 7.1.0.33, and -type truecolor, I get more aliasing than using -colorspace sRGB, which looks fine compared to the IM 6 result.  See the aliasing (stair-stepping) along the top and bottom edges.

**-colorspace sRGB result:**

![test_out2](https://user-images.githubusercontent.com/13458218/167264689-d26d7141-c1dd-4ddf-bb73-97355a906d57.png)

**-type truecolor result:**

![test_out3](https://user-images.githubusercontent.com/13458218/167264709-f8d4c32a-e1eb-4459-a653-c0e2e7dc03f2.png)

**Comment by urban-warrior** [comment]
Use `-type TrueColorAlpha` instead.

**Reply by fmw42** [reply]
OK. Thanks. That works.

**Comment by luco** [comment]
Tested on  7.1.0-40 beta and can confirm that was fixed. Thanks a lot guys for the help!
